packages/framework/presence/src/presenceDatastoreManager.ts

```diff
@@ -195,14 +189,121 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+		data: GeneralDatastoreMessageContent,
+		options: LocalUpdateOptions,
+	): void {
+		const allowableUpdateLatency = options.allowableUpdateLatency ?? 0;
+		const forceBroadcast = options.forceBroadcast;
+
+		const now = Date.now();
+		const updateDeadline = now + allowableUpdateLatency;
+		if (this.queuedData === undefined) {
+			// No queued message, so set the deadline based on the current message's allowable latency
+			this.sendMessageDeadline = updateDeadline;
+		}
+
+		// This function-local "datastore" will hold the merged message data
+		const queueDatastore = this.queuedData ?? {};
+
+		// Iterate over the current message data; individual items are workspaces
+		for (const workspaceName of Object.keys(data)) {
+			const workspaceData = data[workspaceName];
+
+			// Initialize the merged data as the queued datastore entry for the workspace
+			// Since the key might not exist, create an empty object in that case. It will
+			// be set explicitly after the loop.
+			const mergedData = queueDatastore[workspaceName] ?? {};
+
+			// Iterate over each value manager and its data, merging it as needed
+			for (const valueManagerKey of Object.keys(workspaceData)) {
+				for (const [clientSessionId, value] of brandedObjectEntries(
+					workspaceData[valueManagerKey],
+				)) {
+					// TODO: Should any values be ignored here? E.g. values with ignoreUnmonitored?
+					mergedData[valueManagerKey] ??= {};
+					mergedData[valueManagerKey][clientSessionId] = mergeValueDirectory(
+						mergedData[valueManagerKey][clientSessionId],
+						value,
+						0, // TODO: what value should be passed here?
+					);
+					console.log(clientSessionId, value);
+				}
+			}
+
+			// Store the merged data in the function-local queue workspace. The whole contents of this
+			// datstore will be sent as the message data.
+			queueDatastore[workspaceName] = mergedData;
+		}
+
+		if (updateDeadline >= this.sendMessageDeadline) {
+			// Queue the message
+			this.queuedData = queueDatastore;
+			// If the timer has not expired, we can short-circuit because the timer will fire
+			// and cover this update. In other words, queuing this will be fast enough to
+			// meet its deadline, because a timer is already scheduled to fire before its deadline.
+			if (!this.timer.hasExpired() && !forceBroadcast) {
+				return;
+			}
+		}
+
+		// Note that timeoutInMs === allowableUpdateLatency, but the calculation is done this way for clarity.
+		const timeoutInMs = updateDeadline - now;
+		if (timeoutInMs > 0 && !forceBroadcast) {
+			// Schedule the queued messages to be sent at the updateDeadline
+			this.timer.setTimeout(this.sendQueuedMessage.bind(this), timeoutInMs);
+		} else {
+			this.sendQueuedMessage();
+		}
+	}
+
+	/**
+	 * Send any queued signal immediate. Does nothing if no message is queued.
+	 */
+	private sendQueuedMessage(): void {
+		if (this.queuedData === undefined) {
+			return;
+		}
+
+		const clientConnectionId = this.runtime.clientId;
```

Comment:
If not connected here skip the send. Should still clear the queued data.
```typescript
   // Check for connectivity before sending updates.
   if (this.runtime.connected) {
```

Reply:
Won't clearing the queue result in messages being lost? Or is that what we want? Do we have a separate mechanism that handles disconnected messages? I guess we don't want the queue to grow infinitely...

Reply:
Notifications would be lost. States are kept in the datastore; so, when we reconnect all of those are broadcast as part of rejoin.